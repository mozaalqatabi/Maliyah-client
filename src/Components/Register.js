import React, { useEffect, useState } from "react";
import { Form, Button, Alert, Spinner } from "react-bootstrap";
import { useForm } from "react-hook-form";
import { yupResolver } from "@hookform/resolvers/yup";
import { Link, useNavigate } from "react-router-dom";
import { useDispatch, useSelector } from "react-redux";
import { addUser, resetMessage } from "../Features/UserSlice";
import { userSchemaValidation } from "../Validations/UserValidation";

const Register = () => {
  const [name, setName] = useState("");
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [confirmPassword, setConfirmPassword] = useState("");
  const [isSubmitting, setIsSubmitting] = useState(false);
  const [successMsg, setSuccessMsg] = useState("");

  const dispatch = useDispatch();
  const navigate = useNavigate();
  const msg = useSelector((state) => state.counter.message);

  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm({
    resolver: yupResolver(userSchemaValidation),
  });

  const handleFormSubmit = () => {
    const user = { uname: name, password, email };
    setIsSubmitting(true);
    dispatch(addUser(user));
  };

  useEffect(() => {
  if (msg?.toLowerCase().includes("user added")) {
    setSuccessMsg("Signed up successfully!");
    setIsSubmitting(false);

    setTimeout(() => {
      dispatch(resetMessage());

      // ✅ Set onboarding flag
      localStorage.setItem("justRegistered", "true");

      // ✅ Redirect to login, not dashboard
      navigate("/login");
    }, 2000);
  } else {
    setIsSubmitting(false);
  }
}, [msg, navigate, dispatch]);

  return (
    <div
      style={{
        backgroundColor: "#A6BDFF",
        minHeight: "100vh",
        display: "flex",
        justifyContent: "center",
        alignItems: "center",
      }}
    >
      <Link
        to="/"
        style={{
          position: "absolute",
          top: "50px",
          left: "100px",
          color: "#4F6BED",
          textDecoration: "none",
          fontWeight: "500",
          fontSize: "0.95rem",
        }}
      >
        ← Back to Home Page
      </Link>

      <div
        style={{
          width: "100%",
          maxWidth: "420px",
          backgroundColor: "white",
          padding: "30px",
          borderRadius: "12px",
          boxShadow: "0 0 20px rgba(0,0,0,0.1)",
        }}
      >
        <div className="text-center mb-4">
          <h4 className="mt-3 mb-1">Create your Maliyah Account</h4>
          <p style={{ fontSize: "0.9rem", color: "#666" }}>
            Fill in your details to get started
          </p>
        </div>

        <form onSubmit={handleSubmit(handleFormSubmit)}>
          {/* Name Input */}
          <Form.Group controlId="name" className="mb-3">
            <Form.Label>Name</Form.Label>
            <Form.Control
              type="text"
              placeholder="Enter your name"
              {...register("name", {
                value: name,
                onChange: (e) => setName(e.target.value),
              })}
              isInvalid={!!errors.name}
              disabled={isSubmitting}
            />
            <Form.Control.Feedback type="invalid">
              {errors.name?.message}
            </Form.Control.Feedback>
          </Form.Group>

          {/* Email Input (strict email format) */}
          <Form.Group controlId="email" className="mb-3">
            <Form.Label>Email</Form.Label>
            <Form.Control
              type="email"
              placeholder="Enter your email"
              {...register("email", {
                value: email,
                onChange: (e) => setEmail(e.target.value),
              })}
              isInvalid={!!errors.email}
              disabled={isSubmitting}
              pattern="^[^\s@]+@[^\s@]+\.[^\s@]+$"
            />
            <Form.Control.Feedback type="invalid">
              {errors.email?.message || "Please enter a valid email address"}
            </Form.Control.Feedback>
          </Form.Group>

          {/* Password Input */}
          <Form.Group controlId="password" className="mb-3">
            <Form.Label>Password</Form.Label>
            <Form.Control
              type="password"
              placeholder="Enter your password"
              {...register("password", {
                value: password,
                onChange: (e) => setPassword(e.target.value),
              })}
              isInvalid={!!errors.password}
              disabled={isSubmitting}
            />
            <Form.Control.Feedback type="invalid">
              {errors.password?.message}
            </Form.Control.Feedback>
          </Form.Group>

          {/* Confirm Password Input */}
          <Form.Group controlId="confirmPassword" className="mb-3">
            <Form.Label>Confirm Password</Form.Label>
            <Form.Control
              type="password"
              placeholder="Confirm your password"
              {...register("confirmPassword", {
                value: confirmPassword,
                onChange: (e) => setConfirmPassword(e.target.value),
              })}
              isInvalid={!!errors.confirmPassword}
              disabled={isSubmitting}
            />
            <Form.Control.Feedback type="invalid">
              {errors.confirmPassword?.message}
            </Form.Control.Feedback>
          </Form.Group>

          {/* Submit Button */}
          <Button
            className="form-control mb-3"
            type="submit"
            style={{ backgroundColor: "#4F6BED", border: "none" }}
            disabled={isSubmitting}
          >
            {isSubmitting ? (
              <Spinner
                as="span"
                animation="border"
                size="sm"
                role="status"
                aria-hidden="true"
              />
            ) : (
              "Sign Up"
            )}
          </Button>

          {/* Success Message */}
          {successMsg && (
            <Alert variant="success" className="text-center">
              {successMsg}
            </Alert>
          )}

          {/* Error Message */}
          {msg &&
            !msg.toLowerCase().includes("user added") &&
            msg.trim() !== "" && (
              <Alert variant="danger" className="text-center">
                {msg}
              </Alert>
            )}

          {/* Already have an account */}
          <div className="text-center mt-3">
            <small>
              Already have an account?{" "}
              <Link to="/login" style={{ color: "#4F6BED" }}>
                Sign In
              </Link>
            </small>
          </div>
        </form>
      </div>
    </div>
  );
};

export default Register;
