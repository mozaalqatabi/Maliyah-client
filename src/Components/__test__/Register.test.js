import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import Register from '../Register';

const mockStore = configureStore([]);
const store = mockStore({
  counter: { isSuccess: false, isError: false, user: null, message: '' },
});

beforeAll(() => {
  jest.spyOn(console, "warn").mockImplementation(() => {});
});

describe('Register Component Tests', () => {

  test('renders all input fields and submit button', () => {
    render(
      <Provider store={store}>
        <Router>
          <Register />
        </Router>
      </Provider>
    );

    expect(screen.getByLabelText(/name/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/email/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/^password$/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/confirm password/i)).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /sign up/i })).toBeInTheDocument();
  });

  test('allows typing in all input fields', () => {
    render(
      <Provider store={store}>
        <Router>
          <Register />
        </Router>
      </Provider>
    );

    fireEvent.change(screen.getByLabelText(/name/i), { target: { value: 'John Doe' } });
    fireEvent.change(screen.getByLabelText(/email/i), { target: { value: 'john@example.com' } });
    fireEvent.change(screen.getByLabelText(/^password$/i), { target: { value: 'Abc@123' } });
    fireEvent.change(screen.getByLabelText(/confirm password/i), { target: { value: 'Abc@123' } });

    expect(screen.getByLabelText(/name/i).value).toBe('John Doe');
    expect(screen.getByLabelText(/email/i).value).toBe('john@example.com');
    expect(screen.getByLabelText(/^password$/i).value).toBe('Abc@123');
    expect(screen.getByLabelText(/confirm password/i).value).toBe('Abc@123');
  });

  test('validates email format using regex', () => {
    render(
      <Provider store={store}>
        <Router>
          <Register />
        </Router>
      </Provider>
    );

    const emailInput = screen.getByLabelText(/email/i);
    const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;

    fireEvent.change(emailInput, { target: { value: 'valid.email@example.com' } });
    expect(emailRegex.test(emailInput.value)).toBe(true);

    fireEvent.change(emailInput, { target: { value: 'invalid-email' } });
    expect(emailRegex.test(emailInput.value)).toBe(false);
  });

  test('shows validation errors for empty fields on submit', async () => {
    render(
      <Provider store={store}>
        <Router>
          <Register />
        </Router>
      </Provider>
    );

    fireEvent.click(screen.getByRole('button', { name: /sign up/i }));

    expect(await screen.findByText(/please enter your name/i)).toBeInTheDocument();
    expect(await screen.findByText(/please enter your email/i)).toBeInTheDocument();
    expect(await screen.findByText(/please enter your password/i)).toBeInTheDocument();
    expect(await screen.findByText(/please confirm your password/i)).toBeInTheDocument();
  });

  test('matches the UI snapshot', () => {
    const { container } = render(
      <Provider store={store}>
        <Router>
          <Register />
        </Router>
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });

});
