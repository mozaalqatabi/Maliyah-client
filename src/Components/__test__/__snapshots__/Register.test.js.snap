// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register Component Tests matches the UI snapshot 1`] = `
<div>
  <div
    style="background-color: rgb(166, 189, 255); min-height: 100vh; display: flex; justify-content: center; align-items: center;"
  >
    <a
      href="/"
      style="position: absolute; top: 50px; left: 100px; color: rgb(79, 107, 237); text-decoration: none; font-weight: 500; font-size: 0.95rem;"
    >
      ‚Üê Back to Home Page
    </a>
    <div
      style="width: 100%; max-width: 420px; background-color: white; padding: 30px; border-radius: 12px; box-shadow: 0 0 20px rgba(0,0,0,0.1);"
    >
      <div
        class="text-center mb-4"
      >
        <h4
          class="mt-3 mb-1"
        >
          Create your Maliyah Account
        </h4>
        <p
          style="font-size: 0.9rem; color: rgb(102, 102, 102);"
        >
          Fill in your details to get started
        </p>
      </div>
      <form>
        <div
          class="mb-3"
        >
          <label
            class="form-label"
            for="name"
          >
            Name
          </label>
          <input
            class="form-control"
            id="name"
            name="name"
            placeholder="Enter your name"
            type="text"
          />
          <div
            class="invalid-feedback"
          />
        </div>
        <div
          class="mb-3"
        >
          <label
            class="form-label"
            for="email"
          >
            Email
          </label>
          <input
            class="form-control"
            id="email"
            name="email"
            pattern="^[^\\\\s@]+@[^\\\\s@]+\\\\.[^\\\\s@]+$"
            placeholder="Enter your email"
            type="email"
          />
          <div
            class="invalid-feedback"
          >
            Please enter a valid email address
          </div>
        </div>
        <div
          class="mb-3"
        >
          <label
            class="form-label"
            for="password"
          >
            Password
          </label>
          <input
            class="form-control"
            id="password"
            name="password"
            placeholder="Enter your password"
            type="password"
          />
          <div
            class="invalid-feedback"
          />
        </div>
        <div
          class="mb-3"
        >
          <label
            class="form-label"
            for="confirmPassword"
          >
            Confirm Password
          </label>
          <input
            class="form-control"
            id="confirmPassword"
            name="confirmPassword"
            placeholder="Confirm your password"
            type="password"
          />
          <div
            class="invalid-feedback"
          />
        </div>
        <button
          class="form-control mb-3 btn btn-primary"
          style="background-color: rgb(79, 107, 237);"
          type="submit"
        >
          Sign Up
        </button>
        <div
          class="text-center mt-3"
        >
          <small>
            Already have an account?
             
            <a
              href="/login"
              style="color: rgb(79, 107, 237);"
            >
              Sign In
            </a>
          </small>
        </div>
      </form>
    </div>
  </div>
</div>
`;
